alibi/explainers/similarity/grad.py

```diff
@@ -0,0 +1,203 @@
+"""Gradient-based explainer.
+
+This module implements the gradient-based explainers grad-dot and grad-cos.
+"""
+
+import copy
+from typing import TYPE_CHECKING, Callable, Optional, Union, Dict, Tuple
+
+import numpy as np
+
+from alibi.api.interfaces import Explanation
+from alibi.explainers.similarity.base import BaseSimilarityExplainer
+from alibi.explainers.similarity.metrics import dot, cos, asym_dot
+from alibi.api.defaults import DEFAULT_META_SIM, DEFAULT_DATA_SIM
+
+if TYPE_CHECKING:
+    import tensorflow
+    import torch
+
+
+class SimilarityExplainer(BaseSimilarityExplainer):
+    def __init__(self,
+                 predictor: 'Union[tensorflow.keras.Model, torch.nn.Module]',
+                 loss_fn: '''Callable[[Union[tensorflow.Tensor, torch.Tensor],
+                                    Union[tensorflow.Tensor, torch.Tensor]],
+                                   Union[tensorflow.Tensor, torch.Tensor]]''',
+                 sim_fn: 'Union[Callable[[np.ndarray, np.ndarray], np.ndarray], str]' = 'grad_dot',
+                 task: str = "classification",
+                 store_grads: bool = False,
+                 seed: int = 0,
+                 backend: str = "tensorflow",
+                 **kwargs
+                 ):
+        """Constructor
+
+        Parameters
+        ----------
+        predictor:
+            model to explain.
+        loss_fn:
+            Loss function used.
+        sim_fn:
+            Similarity function to use. ``'grad_dot'`` | ``'grad_cos'``. Default: ``'grad_dot'``.
+        task:
+            Task to perform. ``'classification'`` | ``'regression'``. Default: ``'classification'``.
+        store_grads:
+            Whether to store gradients. Default ``False``. If ``False``, gradients are computed on the fly otherwise we
+            store them which can be faster when it comes to explaining and instances.
+        seed:
+            Random seed. Default: 0.
+        backend:
+            Backend to use. ``'tensorflow'`` | ``'torch'``. Default: ``'tensorflow'``.
+        kwargs:
+            Additional arguments to pass to the similarity function.
+        """
+
+        self.meta = copy.deepcopy(DEFAULT_META_SIM)
+        self.meta['params'].update(
+            sim_fn_name=sim_fn,
+            store_grads=store_grads,
+            seed=seed,
+            backend_name=backend,
+            task_name=task
+        )
+
+        sim_fn_opts: Dict[str, Callable] = {
+            'grad_dot': dot,
+            'grad_cos': cos,
+            'grad_asym_dot': asym_dot
+        }
+
+        if isinstance(sim_fn, str):
+            if sim_fn not in sim_fn_opts.keys():
+                raise ValueError(f'Unknown method {sim_fn}. Consider using: `{"` | `".join(sim_fn_opts.keys())}`.')
+
+            sim_fn = sim_fn_opts[sim_fn]
+
+        if task not in ['classification', 'regression']:
```

Comment:
Consider creating a `Task` enum and using it here.

Reply:
So I've created a Task enum but I figure it doesn't make sense to use here as we want the `task` parameter in the public API to be a Literal. After the error checking is performed here I convert to the Task Enum class.